File: Example/QDNavigationBar/QDConfigEditor/QDConfigSegmentView.m
//
//  QDConfigSegmentView.m
//  QDNavigationBar_Example
//
//  Created by Sinno on 2021/1/30.
//  Copyright © 2021 CocoaPods. All rights reserved.
//

#import "QDConfigSegmentView.h"
#import <Masonry/Masonry.h>

@interface QDConfigSegmentView ()
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UISegmentedControl *segmentControl;
@property (nonatomic, strong) NSArray *items;
@end

@implementation QDConfigSegmentView

#pragma mark - Initialization
// 初始化方法

- (instancetype)initWithItems:(NSArray *)items {
    if (self = [super initWithFrame:CGRectZero]) {
        self.items = items;
        [self configSubviews];
    }
    return self;
}

#pragma mark - Super methods

#pragma mark - Action Methods
- (void)onValueChanged {
    if (self.valueChanged) {
        self.valueChanged(self.segmentControl.selectedSegmentIndex);
    }
}

#pragma mark - Public Methods
// 公开的方法
- (void)setTitle:(NSString *)title {
    self.titleLabel.text = title;
}

- (NSString *)title {
    return self.titleLabel.text;
}

- (void)setSelectedIndex:(NSInteger)index {
    self.segmentControl.selectedSegmentIndex = index;
}

- (NSInteger)selectedIndex {
    return self.segmentControl.selectedSegmentIndex;
}

#pragma mark - Private Methods
// 私有的方法
- (void)configSubviews {
    [self addSubview:self.titleLabel];
    [self addSubview:self.segmentControl];

    [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self);
        make.leading.equalTo(self).offset(15);
        make.trailing.lessThanOrEqualTo(self.segmentControl.mas_leading).offset(-15);
    }];
    
    [self.segmentControl mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self);
        make.trailing.equalTo(self).offset(-15);
    }];
}
#pragma mark - Getter && Setter
// Getter和Setter方法
- (UILabel *)titleLabel {
    if (!_titleLabel) {
        UILabel *label = [[UILabel alloc] init];
        label.font = [UIFont systemFontOfSize:20.0];
        label.textColor = UIColor.blackColor;
        label.adjustsFontSizeToFitWidth = YES;
        label.minimumScaleFactor = 0.8;
        _titleLabel = label;
    }
    return _titleLabel;
}

- (UISegmentedControl *)segmentControl {
    if (!_segmentControl) {
        UISegmentedControl *view = [[UISegmentedControl alloc] initWithItems:self.items];
        view.apportionsSegmentWidthsByContent = YES;
        [view addTarget:self action:@selector(onValueChanged) forControlEvents:UIControlEventValueChanged];
        _segmentControl = view;
    }
    return _segmentControl;
}

@end

File: Example/QDNavigationBar/QDConfigEditor/QDConfigSegmentView.swift
//
//  QDConfigSegmentView.swift
//  QDNavigationBar_Example
//
//  Created by zqd on 2024/6/5.
//  Copyright © 2024 CocoaPods. All rights reserved.
//

import UIKit
import SnapKit

class QDConfigSegmentView: UIView {
    
    private var titleLabel: UILabel = {
        let label = UILabel()
        label.font = UIFont.systemFont(ofSize: 20.0)
        label.textColor = .black
        label.adjustsFontSizeToFitWidth = true
        label.minimumScaleFactor = 0.8
        return label
    }()
    
    private var segmentControl: UISegmentedControl!
    private var items: [Any]
    
    var valueChanged: ((Int) -> Void)?
    
    var title: String? {
        get {
            return titleLabel.text
        }
        set {
            titleLabel.text = newValue
        }
    }
    
    var selectedIndex: Int {
        get {
            return segmentControl.selectedSegmentIndex
        }
        set {
            segmentControl.selectedSegmentIndex = newValue
        }
    }
    
    // MARK: - Initialization
    
    init(items: [Any]) {
        self.items = items
        super.init(frame: .zero)
        configSubviews()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: - Private Methods
    
    private func configSubviews() {
        segmentControl = UISegmentedControl(items: items)
        segmentControl.apportionsSegmentWidthsByContent = true
        segmentControl.addTarget(self, action: #selector(onValueChanged), for: .valueChanged)
        
        addSubview(titleLabel)
        addSubview(segmentControl)
        
        titleLabel.snp.makeConstraints { make in
            make.centerY.equalToSuperview()
            make.leading.equalToSuperview().offset(15)
            make.trailing.lessThanOrEqualTo(segmentControl.snp.leading).offset(-15)
        }
        
        segmentControl.snp.makeConstraints { make in
            make.centerY.equalToSuperview()
            make.trailing.equalToSuperview().offset(-15)
        }
    }
    
    @objc private func onValueChanged() {
        valueChanged?(segmentControl.selectedSegmentIndex)
    }
}

